### PR TITLE
Add util.MachineToInfrastructureMapFunc for infrastructure providers

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vincepri@vmware.com>

**What this PR does / why we need it**:
This PR adds a helper function for infrastructure providers to use to watch Machine events and enqueue reconciliation requests for the referenced provider types.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**Release note**:

```release-note

```

### CODE REVIEW

File: pkg/util/util.go [hunk 1]
```diff
@@ -149,6 +152,32 @@ func IsNodeReady(node *v1.Node) bool {
...
+func MachineToInfrastructureMapFunc(gvk schema.GroupVersionKind) handler.ToRequestsFunc {
+	return func(o handler.MapObject) []reconcile.Request {
+		m, ok := o.Object.(*clusterv1.Machine)
+		if !ok {
+			return nil
+		}
+
+		// Return early if the GroupVersionKind don't match what we expect.
+		infraGVK := m.Spec.InfrastructureRef.GroupVersionKind()
+		if !reflect.DeepEqual(gvk, infraGVK) {
```
Comment (ncdc) [inline pkg/util/util.go]: You can use `==` to compare two GVKs

File: pkg/util/util.go [hunk 2]
```diff
@@ -149,6 +152,32 @@ func IsNodeReady(node *v1.Node) bool {
 	return false
 }
 
+// MachineToInfrastructureMapFunc returns a handler.ToRequestsFunc that watches for
+// Machine events and returns reconciliation requests for an infrastructure provider object.
+func MachineToInfrastructureMapFunc(gvk schema.GroupVersionKind) handler.ToRequestsFunc {
```
Comment (ncdc) [inline pkg/util/util.go]: Why in `util` and not in the machine controller code?

**Comment by ncdc** [comment]
Also, could you please add a unit test?

File: pkg/util/util.go (hunk 2 above)
Comment (vincepri) [reply pkg/util/util.go]: It feels weird to import the machine controller code from a provider controller, wdyt?

File: pkg/util/util.go [hunk 3]
```diff
@@ -149,6 +152,32 @@ func IsNodeReady(node *v1.Node) bool {
...
+	return func(o handler.MapObject) []reconcile.Request {
+		m, ok := o.Object.(*clusterv1.Machine)
+		if !ok {
+			return nil
+		}
+
+		// Return early if the GroupVersionKind don't match what we expect.
```
Comment (ncdc) [inline pkg/util/util.go]:
```suggestion
		// Return early if the GroupVersionKind doesn't match what we expect.
```

File: pkg/util/util.go (hunk 2 above)
Comment (ncdc) [reply pkg/util/util.go]: Oh that's right, this is meant to be used by infra providers. It'd be kinda nice to see this in something like `pkg/machine` as `InfrastructureMapFunc` - i.e. `machine.InfrastructureMapFunc`. But I realize having a package named `machine` would conflict with having a `Machine` with the same name.

Long-term, I want to eliminate `pkg/util` and move everything into other packages. `util` is too much of a grab bag.

I suppose it's ok to lump it in here for now 😄

**Comment by vincepri** [comment]
Should be good to go

**Comment by vincepri** [comment]
Had to rebase, @detiber can I get another lgtm here? 😄

**Pull request merged**